### PR TITLE
refactor: use types rather than `Any` in `flightsql` interface

### PR DESCRIPTION
**Pull request by alamb**
~Draft: builds on https://github.com/influxdata/influxdb_iox/pull/6703~

Follow up for @tustvold's suggestion in https://github.com/influxdata/influxdb_iox/pull/6667#discussion_r1086468817

re https://github.com/influxdata/influxdb_iox/issues/6471

# Rationale
the interface is somewhat hard to follow due to the use of Any types that are encoded/decoded. 

# Changes
1. Refactor the interface so it uses specific types rather than the `Any` type and encapsulate the Any conversion in `FlightSQLCommand`

### CODE REVIEW

File: flightsql/src/cmd.rs [hunk 1]
```diff
@@ -59,29 +67,47 @@ pub enum FlightSQLCommand {
     ActionClosePreparedStatementRequest(PreparedStatementHandle),
 }
 
+impl Display for FlightSQLCommand {
```
Comment (alamb) [inline flightsql/src/cmd.rs:70]: This display is used for logging and query_log table

File: service_common/src/planner.rs [hunk 2]
```diff
@@ -70,7 +69,7 @@ impl Planner {
         &self,
         namespace_name: impl Into<String>,
         namespace: Arc<N>,
-        msg: Any,
+        cmd: FlightSQLCommand,
```
Comment (alamb) [inline service_common/src/planner.rs:72]: this is the key change -- don't use Any here and instead use `FlightSQLCommand` that handles the conversion back and forth to bytes following FlightSQL protobuf

**Comment by alamb** [comment]
@tustvold  can I please bother you for a review of this PR (as I made the changes at your suggestion) 🙏

**Comment by alamb** [comment]
Can haz review please @tustvold ?

File: service_grpc_flight/src/lib.rs [hunk 3]
```diff
@@ -634,11 +638,11 @@ where
...
 /// Extracts an encoded Protobuf message from a [`FlightDescriptor`],
 /// as used in FlightSQL.
-fn msg_from_descriptor(flight_descriptor: FlightDescriptor) -> Result<Any> {
+fn cmd_from_descriptor(flight_descriptor: FlightDescriptor) -> Result<FlightSQLCommand> {
     match flight_descriptor.r#type() {
         DescriptorType::Cmd => {
-            let msg: Any = Message::decode(flight_descriptor.cmd).context(DeserializationSnafu)?;
-            Ok(msg)
+            let cmd = FlightSQLCommand::try_decode(flight_descriptor.cmd)?;
+            Ok(cmd)
```
Comment (tustvold) [inline service_grpc_flight/src/lib.rs:645]:
```suggestion
            FlightSQLCommand::try_decode(flight_descriptor.cmd)
```
Perhaps, not sure about the error types

File: flightsql/src/cmd.rs [hunk 4]
```diff
@@ -27,17 +30,22 @@ impl PreparedStatementHandle {
...
         self.query.as_ref()
     }
-}
-
-/// Decode bytes to a PreparedStatementHandle
-impl TryFrom<Bytes> for PreparedStatementHandle {
-    type Error = Error;
 
-    fn try_from(handle: Bytes) -> Result<Self, Self::Error> {
+    fn try_decode(handle: Bytes) -> Result<Self> {
```
Comment (tustvold) [inline flightsql/src/cmd.rs:34]: Could possibly add link to #6699

File: flightsql/src/planner.rs [hunk 5]
```diff
@@ -27,13 +27,12 @@ impl FlightSQLPlanner {
     /// Returns the schema, in Arrow IPC encoded form, for the request in msg.
     pub async fn get_flight_info(
         namespace_name: impl Into<String>,
-        msg: Any,
+        cmd: FlightSQLCommand,
```
Comment (tustvold) [inline flightsql/src/planner.rs:30]:
```suggestion
        cmd: &FlightSQLCommand,
```
Perhaps

File: flightsql/src/planner.rs [hunk 6]
```diff
@@ -71,13 +70,12 @@ impl FlightSQLPlanner {
     pub async fn do_get(
         namespace_name: impl Into<String>,
         _database: Arc<dyn QueryNamespace>,
-        msg: Any,
+        cmd: FlightSQLCommand,
```
Comment (tustvold) [inline flightsql/src/planner.rs:73]:
```suggestion
        cmd: &FlightSQLCommand,
```

File: flightsql/src/planner.rs [hunk 7]
```diff
@@ -102,13 +100,12 @@ impl FlightSQLPlanner {
     pub async fn do_action(
         namespace_name: impl Into<String>,
         _database: Arc<dyn QueryNamespace>,
-        msg: Any,
+        cmd: FlightSQLCommand,
```
Comment (tustvold) [inline flightsql/src/planner.rs:103]:
```suggestion
        cmd: &FlightSQLCommand,
```

File: service_common/src/planner.rs (hunk 2 above)
Comment (tustvold) [inline service_common/src/planner.rs:72]:
```suggestion
        cmd: &FlightSQLCommand,
```

File: service_common/src/planner.rs [hunk 8]
```diff
@@ -117,14 +116,14 @@ impl Planner {
     pub async fn flight_sql_get_flight_info(
         &self,
         namespace_name: impl Into<String>,
-        msg: Any,
+        cmd: FlightSQLCommand,
```
Comment (tustvold) [inline service_common/src/planner.rs:119]:
```suggestion
        cmd: &FlightSQLCommand,
```

File: service_common/src/planner.rs [hunk 9]
```diff
@@ -94,7 +93,7 @@ impl Planner {
         &self,
         namespace_name: impl Into<String>,
         namespace: Arc<N>,
-        msg: Any,
+        cmd: FlightSQLCommand,
```
Comment (tustvold) [inline service_common/src/planner.rs:96]:
```suggestion
        cmd: &FlightSQLCommand,
```

**Review by tustvold — APPROVED** [review]
Sorry for the delay, it wasn't intentional. Just some minor nits

**Comment by alamb** [comment]
> Sorry for the delay, it wasn't intentional. Just some minor nits

No worries! It is not like you have been sitting around doing nothing 😆

File: flightsql/src/planner.rs (hunk 5 above)
Comment (alamb) [reply flightsql/src/planner.rs:30]: When I tried that I get the compiler telling me that the created future captures a lifetime 

```
cd /Users/alamb/Software/influxdb_iox2 && RUST_BACKTRACE=1 CARGO_TARGET_DIR=/Users/alamb/Software/target-iox2 LOG_FILTER=debug TEST_INFLUXDB_JDBC=true TEST_INTEGRATION=1 TEST_INFLUXDB_IOX_CATALOG_DSN=postgresql://postgres@localhost:5432/postgres cargo test --test end_to_end -- query
   Compiling flightsql v0.1.0 (/Users/alamb/Software/influxdb_iox2/flightsql)
error[E0700]: hidden type for `impl Future<Output = std::result::Result<Arc<(dyn ExecutionPlan + 'static)>, error::Error>>` captures lifetime that does not appear in bounds
  --> flightsql/src/planner.rs:75:41
   |
75 |       ) -> Result<Arc<dyn ExecutionPlan>> {
   |  _________________________________________^
76 | |         let namespace_name = namespace_name.into();
77 | |         debug!(%namespace_name, %cmd, "Handling flightsql do_get");
78 | |
...  |
94 | |         }
95 | |     }
   | |_____^
   |
   = note: hidden type `impl Future<Output = std::result::Result<Arc<(dyn ExecutionPlan + 'static)>, error::Error>>` captures lifetime '_#24r
```

File: service_common/src/planner.rs (hunk 2 above)
Comment (alamb) [reply service_common/src/planner.rs:72]: When I tried to use a reference the compiler complains about captured lifetimes :(

File: flightsql/src/planner.rs (hunk 5 above)
Comment (alamb) [reply flightsql/src/planner.rs:30]: I have some other ideas of how to remove cloning -- will do in a separate PR

File: flightsql/src/planner.rs (hunk 6 above)
Comment (alamb) [reply flightsql/src/planner.rs:73]: I tried a bunch of different ways to make the compiler and failed 😢
